source/Irrlicht/COpenGLDriver.cpp

```diff
@@ -1236,6 +1236,56 @@ bool COpenGLDriver::beginScene(bool backBuffer, bool zBuffer, SColor color,
 }
 
 
+bool COpenGLDriver::bindGraphicsPipeline(video::IGPURenderpassIndependentPipeline* _gpipeline)
+{
+    SAuxContext* ctx = getThreadContext_helper(false);
+    if (!ctx)
+        return false;
+
+    COpenGLRenderpassIndependentPipeline* glpipeline = static_cast<COpenGLRenderpassIndependentPipeline*>(_gpipeline);
+    ctx->nextState.pipeline = core::smart_refctd_ptr<COpenGLRenderpassIndependentPipeline>(glpipeline);
+
+    //all code below this line is descriptor set bindings invalidation according to pipeline layout compatibility rules (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#descriptorsets-compatibility)
+    if (!ctx->currentState.pipeline || !_gpipeline)
+    {
+        std::fill(ctx->nextState.descriptorsParams.descSets, ctx->nextState.descriptorsParams.descSets + IGPUPipelineLayout::DESCRIPTOR_SET_COUNT, nullptr);
+        return true;
+    }
+    
+    uint32_t compatibilityLimit = IGPUPipelineLayout::DESCRIPTOR_SET_COUNT;
+    if (ctx->currentState.pipeline)
+    {
+        const IGPUPipelineLayout* nextLayout = _gpipeline->getLayout();
+        compatibilityLimit = ctx->currentState.pipeline->getLayout()->isCompatibleForSet(IGPUPipelineLayout::DESCRIPTOR_SET_COUNT-1u, nextLayout);
+    }
+
+    if (compatibilityLimit == IGPUPipelineLayout::DESCRIPTOR_SET_COUNT)
+        std::fill(ctx->nextState.descriptorsParams.descSets, ctx->nextState.descriptorsParams.descSets + IGPUPipelineLayout::DESCRIPTOR_SET_COUNT, nullptr);
+    else
+    {
+        for (uint32_t i = compatibilityLimit+1u; i < IGPUPipelineLayout::DESCRIPTOR_SET_COUNT; ++i)
+            ctx->nextState.descriptorsParams.descSets[i] = nullptr;
```

Comment:
wait what!?

you got it wrong, Vulkan counts from index 0 not backwards!

loop should be `for (uint32_t i=0u; i<compatibilityLimit; i++)`

Reply:
well now when i think more about this, i think its completely fucked approach. `bindGraphicsPipeline()` shouldnt touch descriptor sets at all, and `bindDescriptorSets()` should get extra param `_pipelineLayout` (just like Vulkan function has). However vulkan spec says 

>When binding a descriptor set (see Descriptor Set Binding) to set number N, if **the previously bound descriptor sets for sets zero through N-1 were all bound using compatible pipeline layouts** (...)

i cant know which pipeline layout was used to bind descriptor set unless we have a requirement that before every call to `bindGraphicsPipeline()` must go *maximum* of 1 call to `bindDescriptorSets()` (then i know that all bound descriptor sets were bound with pipeline layout compatible with layout of currently bound pipeline)
This seems far more tricky than i thought at first look.. and im not sure i understand it correctly

Reply:
> `bindGraphicsPipeline()` shouldnt touch descriptor sets at all

As much as I would like that to be the case, its the behaviour under the Vulkan spec.

> and `bindDescriptorSets()` should get extra param `_pipelineLayout`

That is also a valid approach.

> i cant know which pipeline layout was used to bind descriptor set unless we have a requirement that before every call to bindGraphicsPipeline() must go maximum of 1 call to bindDescriptorSets() (then i know that all bound descriptor sets were bound with pipeline layout compatible with layout of currently bound pipeline)

Well in my current proposal if `bindGraphicsPipeline` call unbinds no longer compatible descriptor sets (but without flushing state) then by the time you come around with `bindDescriptorSets` you are 100% sure that any non-null bound descriptor sets are compatible with the currently bound pipeline.

Its a similar situation to calling something akin to `bindDescriptorSets(....,driver->getCurrentPipeline()->getLayout)` , vulkan only takes the pipeline layout parameter to bind descriptor sets to count the hardware binding indices properly for OpenGL ES 3.2 class hardware.